core/blockchain.go

```diff
@@ -501,13 +499,14 @@ func (bc *BlockChain) HasBlockAndState(hash common.Hash) bool {
 	return err == nil
 }
 
-// GetBlock retrieves a block from the database by hash, caching it if found.
-func (self *BlockChain) GetBlock(hash common.Hash) *types.Block {
+// getBlock retrieves a block from the database by hash and number,
+// caching it if found.
+func (self *BlockChain) getBlock(hash common.Hash, number uint64) *types.Block {
```

Comment:
I think this could be valuable to retain as public. If I know the number of the header from client code, then I can use that too and avoid a block number lookup. Call this one `GetBlock` and the new one `GetBlockByHash` instead. Then it's more obvious that "by hash" needs more computation.